internal/controller/pkg/revision/establisher.go

```diff
@@ -137,39 +137,64 @@ func (e *APIEstablisher) ReleaseObjects(ctx context.Context, parent v1.PackageRe
 	// referenced resources available and rebuilding the status.
 	// In the next reconciliation loop, and we will be able to remove the
 	// control/ownership of the objects using the new status.
-	objRefs := parent.GetObjects()
-	// Stop controlling objects.
-	for _, ref := range objRefs {
-		u := unstructured.Unstructured{}
-		u.SetAPIVersion(ref.APIVersion)
-		u.SetKind(ref.Kind)
-		u.SetName(ref.Name)
-
-		if err := e.client.Get(ctx, types.NamespacedName{Name: u.GetName()}, &u); err != nil {
-			if kerrors.IsNotFound(err) {
-				// This is not expected, but still not an error for relinquishing.
-				continue
+	allObjs := parent.GetObjects()
+	if len(allObjs) == 0 {
+		return nil
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(maxConcurrentEstablishers)
+	for _, ref := range allObjs {
+		ref := ref // Pin the loop variable.
+		g.Go(func() error {
+			u := unstructured.Unstructured{}
+			u.SetAPIVersion(ref.APIVersion)
+			u.SetKind(ref.Kind)
+			u.SetName(ref.Name)
+
+			if err := e.client.Get(ctx, types.NamespacedName{Name: u.GetName()}, &u); err != nil {
+				if kerrors.IsNotFound(err) {
+					// This is not expected, but still not an error for releasing objects.
+					return nil
+				}
+				return errors.Wrapf(err, errFmtGetOwnedObject, u.GetKind(), u.GetName())
 			}
-			return errors.Wrapf(err, errFmtGetOwnedObject, u.GetKind(), u.GetName())
-		}
-		ors := u.GetOwnerReferences()
-		for i := range ors {
-			if ors[i].UID == parent.GetUID() {
-				ors[i].Controller = ptr.To(false)
-				break
+			ors := u.GetOwnerReferences()
+			changed := false
+			for i := range ors {
+				if ors[i].UID == parent.GetUID() {
+					if ors[i].Controller != nil && *ors[i].Controller {
+						ors[i].Controller = ptr.To(false)
+						changed = true
+					}
+					break
+				}
+				// Note(turkenh): What if we cannot find our UID in the owner
+				// references? This is not expected unless another party stripped
+				// out ownerRefs. I believe this is a fairly unlikely scenario,
+				// and we can ignore it for now especially considering that if that
+				// happens active revision or the package itself will still take
+				// over the ownership of such resources.
 			}
-			// Note(turkenh): What if we cannot find our UID in the owner
-			// references? This is not expected unless another party stripped
-			// out ownerRefs. I believe this is a fairly unlikely scenario,
-			// and we can ignore it for now especially considering that if that
-			// happens active revision or the package itself will still take
-			// over the ownership of such resources.
-		}
-		u.SetOwnerReferences(ors)
-		if err := e.client.Update(ctx, &u); err != nil {
-			return errors.Wrapf(err, errFmtUpdateOwnedObject, u.GetKind(), u.GetName())
-		}
+			if changed {
+				u.SetOwnerReferences(ors)
+				if err := e.client.Update(ctx, &u); err != nil {
+					return errors.Wrapf(err, errFmtUpdateOwnedObject, u.GetKind(), u.GetName())
+				}
+			}
+			select {
+			case <-ctx.Done():
```

Comment:
TBH, I try to follow the [existing](https://github.com/crossplane/crossplane/blob/256178492e736055caed433f75014b938addd511/internal/controller/pkg/revision/establisher.go#L212-L331) implementation in the establisher.

But I see your point and agree that it makes more sense.
Just pushed a [commit](https://github.com/crossplane/crossplane/pull/5044/commits/5469e1ff27a1ab2cbf3320149f7bd6d756d78ba9).

Reply:
As we discussed offline, I've moved the `select` to the beginning of the goroutine. 

With this, all remaining goroutines (we set limit as 10) will immediately return with context error instead of making calls to kube API to get the same error.